### PR TITLE
feat: Add Map batch

### PR DESCRIPTION
**Pull request by RohanAshar**
Fixes https://github.com/numaproj/numaflow/issues/1789
Fixes https://github.com/numaproj/numaflow/issues/1846

### CODE REVIEW

File: src/main/java/io/numaproj/numaflow/batchmapper/Service.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/Service.java:72]: With this `onNext` and `onCompleted` combination, we collect ALL the datums to the iterator and then send the batch to the AKKA actor to iterate through. This approach doesn't leverage the power of AKKA where a single datum once received gets immediately passed to the actor to handle, like a GO channel. From a performance perspective, this is the same as passing in an array of datums.

This leads to two questions: 1. does having input as a channel help improve performance in Go SDK? and 2. if the answer to 1 is yes, then we need to properly use AKKA as we did for reduce stream, which requires two user interfaces. `processMessage` and `handleEndOfStream`, which is not as easy to use as the one in this PR where we just pass in a single iterator.

@kohlisid , @yhl25 WDYT?

**Review by KeranYang — COMMENTED** [review]
Partial Review. Thanks @RohanAshar for quickly pick this up and create the PR. 👍

File: pom.xml
Comment (RohanAshar) [reply pom.xml:155]: Currently for anyone with new M3s the classifier is osx_x86_64. For others its osx-aarch_64 and for intel chips this dependency itself is not present. This is something we need to really look at and see how we can make this dynamic.

Comment (KeranYang) [reply pom.xml:155]: Did a quick search, it seems `os-maven-plugin` with `${os.detected.classifier}` works.

```
<project>
  ...
  <build>
    <plugins>
      <plugin>
        <groupId>kr.motd.maven</groupId>
        <artifactId>os-maven-plugin</artifactId>
        <version>1.7.0</version>
        <executions>
          <execution>
            <goals>
              <goal>detect</goal>
            </goals>
            <phase>initialize</phase>
          </execution>
        </executions>
      </plugin>
      ...
    </plugins>
  </build>

  <dependencies>
    <dependency>
      <groupId>your.group</groupId>
      <artifactId>your-artifact</artifactId>
      <version>your-version</version>
      <classifier>${os.detected.classifier}</classifier>
    </dependency>
    ...
  </dependencies>
  ...
</project>
```

We can test it out.

Comment (RohanAshar) [reply pom.xml:155]: I think this will break the Intel chip version as the dependency is not available for them so the compile would fail. Il revert the changes for now and see if there is an elegant solution in the future.

File: src/main/java/io/numaproj/numaflow/batchmapper/Service.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/Service.java:72]: This is a valid point but the current implementation is not completely async. AFAIK even in Go the sdk gets a batch response which i feel would be the long poll. @kohlisid @yhl25 @vigith Thoughts?

File: src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java:6]: Is the @link annotation working? I see one `{` and two `}`.

File: src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java:19]: blocking queue is the implementation detail of the interface hence should not be exposed in the method description.

```suggestion
     * Returns the next element in the iterator
     * This method blocks until an element becomes available.
     * When EOF_DATUM is received, this method will return null and the iterator will be closed.
     *
     * @return the next element in the iterator, null if EOF_DATUM is received or the iterator is already closed
     *
     * @throws InterruptedException if the thread is interrupted while waiting for the next element
     */
    Datum next() throws InterruptedException;
```

Comment (KeranYang) [reply src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java:19]: Please also update it for sinker interface.

File: src/main/java/io/numaproj/numaflow/batchmapper/Server.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/Server.java:15]:
```suggestion
 * Server is the gRPC server for executing batch map operation.
```

Please also update it for mapstreamer.

File: src/main/java/io/numaproj/numaflow/batchmapper/Service.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/Service.java:82]: Consider adding the real numbers in the error message, for better debuggability.

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java
Comment (KeranYang) [inline src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java:64]:
```suggestion
    public void testBatchMapHappyPath() {
```

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java
Comment (KeranYang) [inline src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java:85]: We are waiting until the size gets to 10 and then verify the size is 10, which is a bit redundant. Can we use `while (!outputStreamObserver.completed.get()) ;` like we did in sinker server test?

Comment (KeranYang) [reply src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java:85]: Can we also verify the content of the response in addition to the size?

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java
Comment (KeranYang) [inline src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java:108]: Is it possible to test the error case where the response count doesn't match the request count?

**Review by KeranYang — COMMENTED** [review]
@RohanAshar one more comment and this is probably an unknown unknown to you - we have a build-push workflow, which automatically builds the example image using the latest change and push it to quay.io. Most of our examples are built into an image and used by numaflow e2e tests, having the auto image update help us ensure the SDK changes are validated by numaflow e2e. As we are going to use the example in the PR for batch map java e2e test, we should add it to the workflow yaml [file](https://github.com/numaproj/numaflow-java/blob/main/.github/workflows/build-push.yaml#L18). https://github.com/numaproj/numaflow-java/blob/main/development.md#adding-a-new-example

File: src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java]: Done

File: src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java]: Done

File: src/main/java/io/numaproj/numaflow/batchmapper/BatchResponse.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/BatchResponse.java:11]: done

File: src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java]: done

File: src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/DatumIterator.java]: done

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java
Comment (RohanAshar) [reply src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java]: done

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java
Comment (RohanAshar) [reply src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java:108]: Added another test case to test that scenario

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java
Comment (RohanAshar) [reply src/test/java/io/numaproj/numaflow/batchmapper/ServerTest.java:85]: Added more checks for content.

File: src/main/java/io/numaproj/numaflow/batchmapper/Service.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/Service.java]: Done

File: src/main/java/io/numaproj/numaflow/batchmapper/Server.java
Comment (RohanAshar) [reply src/main/java/io/numaproj/numaflow/batchmapper/Server.java]: Done

File: src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java
Comment (KeranYang) [reply src/test/java/io/numaproj/numaflow/batchmapper/ServerErrTest.java:108]: 👍

File: src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/batchmapper/BatchMapper.java:11]:
```suggestion
     * method which will be used for processing messages. Please implement the interface to ensure that each message generates a corresponding BatchResponse object with a matching ID.
```

**Review by KeranYang — APPROVED** [review]
LGTM. I am approving with one minor comment. Please monitor the build-push workflow and enable e2e test on numaflow repo. Thanks!

**Comment by KeranYang** [comment]
@RohanAshar I have manually created the repository https://quay.io/repository/numaio/numaflow-java/batch-map-example and granted the CI account permission to push the image. Should be good to monitor the build-push workflow after merging.

**Pull request merged**
